caom2utils/caom2utils/fits2caom2.py

```diff
@@ -1037,7 +1041,203 @@ def get_configed_axes_count(self):
+            'Artifact.productType', index=0, current=artifact.product_type))
+        artifact.release_type = self._to_release_type(self._get_from_list(
+            'Artifact.releaseType', index=0, current=artifact.release_type))
+        artifact.content_type = self._get_from_list(
+            'Artifact.contentType', index=0, current=artifact.content_type)
+        artifact.content_length = self._get_from_list(
+            'Artifact.contentLength', index=0, current=artifact.content_length)
+        artifact.content_checksum = _to_checksum_uri(self._get_from_list(
+            'Artifact.contentChecksum', index=0,
+            current=artifact.content_checksum))
+
+    def _get_set_from_list(self, lookup, index):
+        value = None
+        keywords = None
+        try:
+            keywords = self.blueprint._get(lookup)
+        except KeyError:
+            self.add_error(lookup, sys.exc_info()[1])
+            self.logger.debug(
+                'Could not find \'{}\' in fits2caom2 configuration.'.format(
+                    lookup))
+        if keywords:
+            value = keywords
+            self.logger.debug('{}: assigned value {}.'.format(lookup, value))
+
+        return value
+
+    def _get_from_list(self, lookup, index, current=None):
+        value = None
+        try:
+            keywords = self.blueprint._get(lookup)
+        except KeyError:
+            self.add_error(lookup, sys.exc_info()[1])
+            self.logger.debug(
+                'Could not find {!r} in fits2caom2 configuration.'.format(
+                    lookup))
+            if current:
+                self.logger.debug(
+                    '{}: using current value of {!r}.'.format(lookup, current))
+                value = current
+            return value
+        if keywords:
+            value = keywords
+        elif current:
+            value = current
+
+        self.logger.debug('{}: value is {}'.format(lookup, value))
+        return value
+
+    def add_error(self, key, message):
+        self._errors.append(('{} {} {}'.format(
+            datetime.now().strftime('%Y-%m-%dT%H:%M:%S'), key, message)))
+
+    def _to_data_product_type(self, value):
+        return self._to_enum_type(value, DataProductType)
+
+    def _to_calibration_level(self, value):
+        return self._to_enum_type(value, CalibrationLevel)
+
+    def _to_product_type(self, value):
+        return self._to_enum_type(value, ProductType)
+
+    def _to_release_type(self, value):
+        return self._to_enum_type(value, ReleaseType)
+
+    def _to_enum_type(self, value, to_enum_type):
+        if value is None:
+            raise ValueError(
+                'Must set a value of {} for {}.'.format(to_enum_type.__name__,
+                                                        self.logging_name))
+        elif isinstance(value, to_enum_type):
+            return value
+        else:
+            return to_enum_type(value)
+
+
+class GenericParser(Parser):
```

Comment:
It seems to me that this class does the same thing as Parser. I would make a GenericParser the base non-abstract class that the FitsParser customizes. Postpone the creation of abstract Parser for the future only if needed by other types of parsers.

Reply:
Parser class removed, GenericParser a base non-abstract class.